linkerd/identity/src/lib.rs

```diff
@@ -4,47 +4,71 @@
 mod credentials;
 
 pub use self::credentials::{Credentials, DerX509};
+use linkerd_error::Result;
+use thiserror::Error;
+
+const SPIFFE_URI_SCHEME: &str = "spiffe://";
 
 /// An endpoint identity descriptor used for authentication.
 ///
-/// Practically speaking, this is a DNS-like identity string that matches an
-/// x509 DNS SAN. This will in the future be updated to support SPIFFE IDs as
-/// well.
+/// Practically speaking, this could be:
+/// - a DNS-like identity string that matches an x509 DNS SAN
+/// - an identity in URI form, that matches an x509 URI SAN
 ///
 /// This isn't restricted to TLS or x509 uses. An authenticated Id could be
 /// provided by, e.g., a JWT.
 #[derive(Clone, Debug, Eq, PartialEq, Hash)]
-pub struct Id(pub linkerd_dns_name::Name);
+pub enum Id {
+    Dns(linkerd_dns_name::Name),
+    Uri(http::Uri),
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Error)]
+#[error("invalid TLS id")]
+pub struct InvalidId;
 
 // === impl Id ===
 
 impl std::str::FromStr for Id {
     type Err = linkerd_error::Error;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        // TODO support spiffe:// URIs.
+        if s.starts_with(SPIFFE_URI_SCHEME) {
```

Comment:
Not too sure about this logic here. If we do no want to be SPIFFE specific, should we determine whether it is URI or DNS based on regexes?